Add validation functionality

wagner-intevation opened the pull request:
Adds parameter `--validate` that supports https://github.com/secvisogram/csaf-validator-service to validate the resulting document.

Fixes https://github.com/csaf-tools/CVRF-CSAF-Converter/issues/124

Result examples:

```
$ cvrf2csaf --validate --input-file examples/1.2/cvrf_example_a.xml --force
[...]
2025-05-29 09:25:11,415 - cvrf2csaf - INFO - Valdidation successful.
[...]

$cvrf2csaf --validate --input-file examples/1.2/cvrf_example_d.xml --force
[...]
2025-05-29 09:24:10,392 - cvrf2csaf - WARNING - Some errors occurred during conversion, but producing output as --force option is used.
[...]
2025-05-29 09:24:10,594 - cvrf2csaf - WARNING - Some errors were found at validation: [{'errors': [{'instancePath': '/document/tracking/revision_history/0/number', 'message': 'must match pattern "^(0|[1-9][0-9]*)$|^((0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?)$"'}, {'instancePath': '/document/tracking/revision_history/1/number', 'message': 'must match pattern "^(0|[1-9][0-9]*)$|^((0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?)$"'}], 'warnings': [], 'infos': [], 'isValid': False, 'name': 'csaf_2_0'}, {'errors': [{'instancePath': '/document/tracking/revision_history/0/number', 'message': 'must match pattern "^(0|[1-9][0-9]*)$|^((0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?)$"'}, {'instancePath': '/document/tracking/revision_history/1/number', 'message': 'must match pattern "^(0|[1-9][0-9]*)$|^((0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?)$"'}], 'warnings': [], 'infos': [], 'isValid': False, 'name': 'csaf_2_0_strict'}]
2025-05-29 09:24:10,595 - utils - INFO - Successfully wrote ./id0_invalid.json.
[...]
```

wagner-intevation committed:
Fix running the tests from other directories

Running `pytest` (or calling the test file itseld) in the top directory now can run the tests
or from within the `tests` directory.

Previously, the working directory had to be exactly `tests/test_cvrf_full`.

wagner-intevation committed:
add a validator module using secvisogram/csaf-validator-service

https://github.com/csaf-tools/CVRF-CSAF-Converter/issues/124

wagner-intevation committed:
add validation function to cvrf2csaf and documentation

fixes https://github.com/csaf-tools/CVRF-CSAF-Converter/issues/124

bernhardreiter commented:
Thanks!

> The remaining pylint error is not related to the PR:
> ```
> cvrf2csaf/cvrf2csaf.py:254:0: R0915: Too many statements (52/50)
> (too-many-statements) ```

Can you also deal with it in another PR?
(Silence or rewrite.)

wagner-intevation commented:
> A failure in the mandatory tests from the validator shall make the converter fail I think.

That's what the code in https://github.com/csaf-tools/CVRF-CSAF-Converter/pull/132/files#diff-fa1a64ee695c90205cb0dc8aba4aa7a0dc8eb5e70ede667636ade192a8ddaaa1R359 does

> A question is if validation should be the default

That also means any user of the converter also needs to setup the validator first

bernhardreiter commented:
> That also means any user of the converter also needs to setup the validator first

Yes, this is good. Please add this prerequisite to the documentation.
And at least warn if this validator is missing or cannot be called.
(Or throw an error, unless a flag like --force is given or so. We want to prevent too many options.)

If an error from the validator leads to an overall error from the converter, that is fine (I haven't tested this yet).

wagner-intevation commented:
The validator is now required by default and can be disabled by a parameter. (inverse to previous behavior)

> Yes, this is good. Please add this prerequisite to the documentation.

Added a paragraph on the requirement to the section "Getting started"

> (Or throw an error, unless a flag like --force is given or so. We want to prevent too many options.)

As before, `--force` still allows to ignore validation errors.

I also added handling of http errors (hiding the traceback for e.g. connection failures) and lowered the logging level for the `httpx` class

wagner-intevation committed:
split main function to make pylint happy

solves this pylint error:
cvrf2csaf/cvrf2csaf.py:254:0: R0915: Too many statements (52/50) (too-many-statements)
by splitting the main() function into two

wagner-intevation commented:
As the pylint error (too many statements in `main()`) only appears in this branch, not in branch `main`, I added a fix for it in this PR, in a separate commit, instead of a separate PR.

This PR is now finally ready for a review.

wagner-intevation committed:
validation: implement review feedback

to circumvent pylint error 'Too many branches' in
cvrf2csaf/cvrf2csaf.py, break out one logic block

wagner-intevation commented:
@bernhardreiter this PR is waiting for your review (fixes #124 which blocks #129)

bernhardreiter commented:
`pip install .` does not install the runtime requirements for me (tested with pip 23 and 26). Which breaks the installation instructions.

 I had to add them to setup.cfg

bernhardreiter commented on cvrf2csaf/cvrf2csaf.py:
```diff
@@ -297,8 +299,26 @@ def main():
                              " derived from other sources. Default value is '3.0'.")
 
+    # Validation
+    parser.add_argument('--no-validate', action='store_true',
```
Would it be better spelling to call the option `--no-validation` or `--skip-validation`, or `--skip-validator`?  

As there is no `--validate` option, the opposite reads strange in English.

bernhardreiter commented on cvrf2csaf/cvrf2csaf.py:
```diff
@@ -338,6 +358,15 @@ def main():
+            logging.warning("Some errors were found at validation: %s", validation_result[1])
+        else:
+            logging.info("CSAF validation successful.")
+
```
Would be good to log if the validation was skipped.

wagner-intevation commented:
> `pip install .` does not install the runtime requirements for me (tested with pip 23 and 26). Which breaks the installation instructions.
> 
> I had to add them to setup.cfg


True, `attrs` and `httpx` are present in `pyproject.toml` but not in `setup.cfg`. The latter could be migrated to the first, but that's a different topic. I added the dependencies to the `setup.cfg` file too for compatibility.

wagner-intevation commented:
The tests use Python 3.13, which no longer includes `pkg_resources`. That's not a problem created by this PR.

bernhardreiter commented:
> The tests use Python 3.13, which no longer includes `pkg_resources`. That's not a problem created by this PR.

Can you fix it in a different MR then? :)

wagner-intevation commented:
> > The tests use Python 3.13, which no longer includes `pkg_resources`. That's not a problem created by this PR.
> 
> Can you fix it in a different MR then? :)

→ #134

bernhardreiter commented on cvrf2csaf/cvrf2csaf.py:
```diff
@@ -338,6 +364,17 @@ def main():
+        validation_result = validator.validate(final_csaf)
+        if not validation_result[0]:
+            valid_output = False
+            logging.warning("Some errors were found at validation: %s", validation_result[1])
```
Shouldn't a validation error be an error in the conversion?

bernhardreiter replied:
I think we should abort the conversion if the validation fails, unless --force is given (and the result marked as invalid in the filename).

wagner-intevation replied:
Changed.

IIRC this was on purpose based on discussions.

bernhardreiter commented:
We can now merge main in ...

wagner-intevation commented:
The branch is now rebased on main and the tests succeed now.

reiterl reviewed (dismissed):
LGTM

bernhardreiter commented:
> The branch is now rebased on main and the tests succeed now.

BTW: merging might have been better than rebasing, as I somehow cannot pull forward on the checked out branch.

wagner-intevation commented:
> BTW: merging might have been better than rebasing, as I somehow cannot pull forward on the checked out branch.

`git pull --rebase`

bernhardreiter commented on cvrf2csaf/validate.py:
```diff
@@ -0,0 +1,80 @@
+SECVISOGRAM_TEMPLATE = {
+  "tests": [
+    {
+      "name": "csaf_2_0",
```
This has the example input, but the real input should be different. See http://localhost:8082/docs#/default/post_api_v1_validate on a running secvisogram/csaf-validator-service. 

We should have a way to specify with group of tests are run. The three presets I know of are mandatory, optional, informative, but I do not understand how the api really works.

As for configuration, we can use the same conventions that gocsaf/csaf and isduba use. Here are some links, see validator_presets:
https://github.com/gocsaf/csaf/blob/main/docs/csaf_checker.md
https://github.com/ISDuBA/ISDuBA/blob/main/docs/isdubad-config.md#section_remote_validator

bernhardreiter replied:
I've also opened https://github.com/secvisogram/csaf-validator-service/issues/229 to give feedback about my reading of the documentation.

wagner-intevation replied:
I made the presets configurable via a command-line parameter.

Also, the result of the logging is now easier readable.

bernhardreiter commented:
For testing and debugging, is there are way from the options to see the full result of the validator?

To do it manually, e.g. 
```
apt-get install socat
# in one terminal
socat -v TCP-LISTEN:5051 TCP:localhot:5052
# in the other
 cvrf2csaf --input-file examples/1.2/cvrf_example_e.xml  --force --validator-endpoint http://localhost:5082/api/v1/validate
[..]
2026-05-06 10:31:16,368 - cvrf2csaf - ERROR - Mandatory test is_valid_defined_product_ids failed.
2026-05-06 10:31:16,368 - cvrf2csaf - WARNING - Some errors occurred during conversion, but producing output as --force option is used.
2026-05-06 10:31:16,440 - cvrf2csaf - INFO - CSAF validation successful.

2026-05-06 10:31:16,441 - utils - INFO - Successfully wrote ./acme-2017-42_invalid.json.
```

This shows that on this occasion two tests were run, successfully.
```
{"isValid":true,"tests":[{"errors":[],"warnings":[],"infos":[],"isValid":true,"name":"csaf_2_0"},{"errors":[],"warnings":[],"infos":[],"isValid":true,"name":"csaf_2_0_strict"}]}
```

However I was expecting some failures, secvisogram online, on `acme-2017-42_invalid.json` tells me the error: _/document/tracking/version: version does not match latest revision_  which is correct. So something is  amiss.

When using
```diff
diff --git a/cvrf2csaf/validate.py b/cvrf2csaf/validate.py
index 5de8fe1..cba5b0a 100644
--- a/cvrf2csaf/validate.py
+++ b/cvrf2csaf/validate.py
@@ -17,3 +17,3 @@ SECVISOGRAM_TEMPLATE = {
     {
-      "name": "schema",
+      "name": "mandatory",
       "type": "preset"
``` 

I get the warning I was expecting:
```
2026-05-06 10:53:19,723 - cvrf2csaf - WARNING - Some errors were found at validation: [{'errors': [{'instancePath': '/document/tracking/version', 'message': 'version does not match latest revision'}], 'warnings': [], 'infos': [], 'isValid': False, 'name': 'mandatoryTest_6_1_16'}]
```

We should react to the response accordingly, if isValid is false, we should abort (unless --force is given) with the given errors. Otherwise relay warnings and infos accordingly.

bernhardreiter reviewed (changes requested):
not working as expected (see pending comments.)

wagner-intevation committed:
validation: treat validation errors as errors, unless --force is given

wagner-intevation committed:
validation: configurable presets, pretty output

make the validation presets configurable
make the validation output nicely readable

wagner-intevation committed:
style: work around pylint limit too-many-instance-attributes

bernhardreiter reviewed (changes requested):
When running this with the validator library stopped, I get

```
Traceback (most recent call last):
  File "/home/ber/CVRF-CSAF-Converter.git/venv/bin/cvrf2csaf", line 8, in <module>
    sys.exit(main())
             ~~~~^^
  File "/home/ber/CVRF-CSAF-Converter.git/venv/lib/python3.13/site-packages/cvrf2csaf/cvrf2csaf.py", line 383, in main
    validator.log_result(validation_result[1], logging)
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ber/CVRF-CSAF-Converter.git/venv/lib/python3.13/site-packages/cvrf2csaf/validate.py", line 88, in log_result
    for message in test[f"{log_t}s"]:
                   ~~~~^^^^^^^^^^^^^
TypeError: string indices must be integers, not 'str'
```

adding a print shows
```
validation_result = (False, '[Errno 111] Connection refused')
```

so there is no test classification  for cvrf2csaf/validate.py to be found.

bernhardreiter committed:
improve validator service call

 * change default preset to `basic` which includes the mandatory tests,
   the mandatory schema check, and the SHOULD not have
   additionalProperties test.
 * add new option to README

wagner-intevation committed:
fix: handle validation service errors

bernhardreiter committed:
Merge branch 'main' into 124-validate

bernhardreiter committed:
fix usage message for validator-preset default

bernhardreiter reviewed (dismissed):
LGTM & works

wagner-intevation committed:
ci: install setuptools

setuptools now needs to be installed explicitly

https://github.com/csaf-tools/CVRF-CSAF-Converter/pull/132#issuecomment-5291782441

bernhardreiter reviewed (approved):
LGTM

reiterl reviewed (approved):
LGTM & tests of some example xmls looks okay ( not throwing exceptions)

I have started a validator and tested against the 1.2 example files with/without force.
I saw different outputs, but no traceback.
So okay.
